python/ray/monitor.py

```diff
@@ -283,16 +294,86 @@ def plasma_manager_heartbeat_handler(self, channel, data):
         # manager.
         self.live_plasma_managers[db_client_id] = 0
 
-    def driver_removed_handler(self, channel, data):
+    def _cleanup_entries_for_driver(self, driver_id, redis_shard_id):
+        """Remove this driver's objects/tasks entries from a redis shard."""
+        # TODO(zongheng): consider adding save & restore functionalities.
+        # TODO(zongheng): handle function_table, client_table, log_files --
+        # these are in the metadata redis server, not in the shards.
+
+        redis = self.state.redis_clients[redis_shard_id]
+        task_table_infos = {}  # task id -> TaskInfo messages
+
+        # Scan the task table & filter to get the list of tasks belong to this
+        # driver.  Use a cursor in order not to block the redis shards.
+        cursor = "0"
+        while cursor != 0:
+            cursor, keys = redis.scan(
+                cursor=cursor, match=TASK_TABLE_PREFIX + "*")
+            for key in keys:
+                entry = redis.hgetall(key)
+                task_info = TaskInfo.GetRootAsTaskInfo(entry["TaskSpec"], 0)
+                if driver_id != task_info.DriverId():
+                    # Ignore tasks that aren't from this driver.
+                    continue
+                task_table_infos[task_info.TaskId()] = task_info
+
+        # Get the list of objects returned by these tasks.
+        binary_object_ids = []
+        for task_info in task_table_infos.values():
+            binary_object_ids.extend([
+                task_info.Returns(i) for i in range(task_info.ReturnsLength())
+            ])
+
+        # Also record the objects ray.put()'d from the driver.
+        relevant_task_ids = set(task_table_infos.keys())
+        cursor = "0"
+        while cursor != 0:
+            cursor, keys = redis.scan(
+                cursor=cursor, match=OBJECT_INFO_PREFIX + "*")
+            for key in keys:
+                entry = redis.hgetall(key)
+                if entry["is_put"] == "0":
+                    continue
+                parsed_task_id = entry["task"]
+                if parsed_task_id in relevant_task_ids:
+                    binary_object_id = key.split(OBJECT_INFO_PREFIX)[1]
+                    binary_object_ids.append(binary_object_id)
+
+        # Save entries for objects.
+        object_ids_locs = set()
+        object_ids_infos = set()
+        for object_id in binary_object_ids:
+            # OL.
+            obj_loc = redis.zrange(OBJECT_LOCATION_PREFIX + object_id, 0, -1)
```

Comment:
Great question and you're right.  However, I intentionally wrote it this way because in the future, we want to persist the returned `obj_loc`s and `obj_info`s.  If they were empty, we presumably want to special-case their serde somehow (either short-circuit, or write a special dummy value, something like that).  For now it doesn't matter either way.  

Let me know if you still want me to re-write it in the simpler way?